### PR TITLE
Add lints for digit grouping.

### PR DESCRIPTION
**Pull request by jayhardee9**
Issue #1736 
Checks for:
* long literals that don't have any underscores,
* literals that do, but are used inconsistently (like 1_234_5678), and
* literals that have groups larger than 4 digits (like 0b10110001_10001101).
It seems that I have updated some stderr files as well.

### CODE REVIEW

**Comment by clarcharr** [comment]
counterexamples to "inconsistent" literals: `1_000_00` (one thousand dollars, two cents)

I'd prefer to not lint "inconsistent" grouping because it isn't able to predict why a user is grouping things

File: clippy_lints/src/literal_digit_grouping.rs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+impl EarlyLintPass for LiteralDigitGrouping {
+    fn check_expr(&mut self, cx: &EarlyContext, expr: &Expr) {
+        if in_external_macro(cx, expr.span) {
+            return;
+        }
+
+        match expr.node {
+            ExprKind::Lit(ref lit) => self.check_lit(cx, lit),
+            _ => (),
```
Comment (mcarton) [inline clippy_lints/src/literal_digit_grouping.rs]: Use `if let` here.

File: clippy_tests/examples/len_zero.stderr [hunk 2]
```diff
@@ -46,45 +46,45 @@ error: length comparison to zero
    --> len_zero.rs:130:8
     |
 130 |     if x.len() == 0 {
-    |        ^^^^^^^^^^^^ help: using `is_empty` is more concise: `x.is_empty()`
+    |        ^^^^^^^^^^^^ help: using `is_empty` is more concise:: `x.is_empty()`
```
Comment (mcarton) [inline clippy_tests/examples/len_zero.stderr]: Did `rustc` add a colon here?

File: clippy_lints/src/literal_digit_grouping.rs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+/// ```rust
+/// 61864918973511
+/// ```
+declare_lint! {
+    pub UNREADABLE_LITERAL,
+    Warn,
+    "long integer literal without underscores"
+}
+
+/// **What it does:** Warns if an integral constant is grouped inconsistently with underscores.
```
Comment (mcarton) [inline clippy_lints/src/literal_digit_grouping.rs]: The lint also work on floating point constant.

File: clippy_lints/src/literal_digit_grouping.rs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+/// ```rust
+/// 618_64_9189_73_511
+/// ```
+declare_lint! {
+    pub INCONSISTENT_DIGIT_GROUPING,
+    Warn,
+    "integer literals with digits grouped inconsistently"
+}
+
+/// **What it does:** Warns if the digits of an integral constant are grouped into groups that
```
Comment (mcarton) [inline clippy_lints/src/literal_digit_grouping.rs]: The lint also work on floating point constant.

File: clippy_lints/src/literal_digit_grouping.rs [hunk 5]
```diff
@@ -0,0 +1,213 @@
+//! Lints concerned with the grouping of digits with underscores in integral or
+//! floating-point literal expressions.
+
+use rustc::lint::*;
+use syntax::ast::*;
+use syntax_pos;
+use utils::{span_help_and_lint, snippet_opt, in_external_macro};
+
+/// **What it does:** Warns if a long integral constant does not contain underscores.
```
Comment (mcarton) [inline clippy_lints/src/literal_digit_grouping.rs]: The lint also work on floating point constant.

**Comment by clarcharr** [comment]
@mcarton I understand, but that is a bit verbose and personally I'd like to see it the other way. And as I said in the issue thread, there are number systems which have numbering that doesn't follow this rule; see https://en.wikipedia.org/wiki/Decimal_mark#Digit_grouping

File: clippy_tests/examples/len_zero.stderr (hunk 2 above)
Comment (jayhardee9) [inline clippy_tests/examples/len_zero.stderr]: I believe so. Removing the colon from https://github.com/rust-lang-nursery/rust-clippy/blob/master/clippy_lints/src/len_zero.rs#L176 gets rid of the duplicate.

**Comment by jayhardee9** [comment]
@mcarton Sure, I will add suggestions!
@clarcharr I think groupings like that are definitely valid in some situations, but in most cases, if a programmer types something like "1_000_00", I'd bet they actually made a typo.

**Comment by mcarton** [comment]
I agree that `1_000_00` should be linted about. It would be a typo in the huge majority of projects. And I know that languages have different rules of how to write number, but in Rust, the convention is English. If I saw `1.234` in my language, I would read "one thousand two hundred and thirty four", but if I read this in Rust, in know it is "one and two hundred and thirty four thousandth". The few projects which would use other conventions could always `#[allow]` the lint.

**Comment by oli-obk** [comment]
The ui tests also need to be updated: `tests/ui/update-all-references.sh clippy_tests/target/debug/test_build_base`

**Comment by jayhardee9** [comment]
@oli-obk I ran that script, and nothing happened. Is that normal?

**Comment by oli-obk** [comment]
> I ran that script, and nothing happened. Is that normal?

Yea, after a rebase nothing should happen anymore :)

I quickly fixed a small dogfood issue. Let's see if we can get your PR into today's publish.

**Comment by oli-obk** [comment]
> Let's see if we can get your PR into today's publish.

Jup, works now.

So. I like the impl and think that the current defaults are sensible. We should fine-tune it if/once we get complaints. Maybe even make it configurable.

**Pull request merged**